### PR TITLE
models: add original dBFT 2.0 specifications

### PR DESCRIPTION
**Pull request by AnnaShaleva**
An essential part of https://github.com/nspcc-dev/neo-go/issues/2552 investigation.

TODO:

- [x] Finish model description, documentation and usage notes.
- [x] Add model schemas.
- [x] Add model checker configurations.
- [x] ~~Attach error traces for known deadlocks.~~ Will be done in a separate issue.
- [x] Add .gitignore for TLC files.

### CODE REVIEW

File: formal-models/dbft/dbft.tla [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+        /\ msgs' = msgs \cup cv
+        /\ UNCHANGED <<rmState>>
+
+\* RMFaultyDoCV describes view changing by the faulty node r.
+RMFaultyDoCV(r) ==
+  /\ rmState[r].type = "bad"
+  \* TODO: why do we have to check CV not in msgs?
+  /\ [type |-> "ChangeView", rm |-> r, view |-> rmState[r].view] \notin msgs
+  /\ rmState' = [rmState EXCEPT ![r].view = GetNewView(rmState[r].view)]
+  /\ UNCHANGED <<msgs>>
```
Comment (AnnaShaleva) [inline formal-models/dbft/dbft.tla:220]: @roman-khimov, could you, please, explain, what was the idea of adding `  /\ [type |-> "ChangeView", rm |-> r, view |-> rmState[r].view] \notin msgs` line to this action? IMO, it is not needed here, faulty node can do CV whenever it wants.

File: formal-models/README.md [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+decision in a few consensus rounds (views). Each consensus node has its own state
+at each step of the behaviour. Consensus node may send a consensus message by
+adding it to the `msgs` pool. To perform the transition between states the
+consensus node must send a consensus message or there must be a particular set of
+consensus messages in the shared message pool required for a particular
+transition.
+
+Here's the scheme of transitions between consensus node states:
+
+![Basic dBFT model transitions scheme](./.github/dbft.png)
```
Comment (roman-khimov) [inline formal-models/README.md:37]: Do we have some editable source for it?

File: formal-models/README.md [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+[basic model](#basic-dbft-20-model) in the following way: besides the shared pool
+of consensus messages `msgs` each consensus node has its own local pool of
+received and handled messages. Decisions on transmission between the node states
+are taken by the node based on the state of the local message pool. This approach
+allows to create more accurate low-leveled model which is extremely close to the
+dBFT implementation presented in this repository. At the same time such approach
+*significantly* increases the number of considered model states which leads to
+abnormally long TLC Model Checker runs. Thus, we do not recommend to use this
+model in development and place it here as an example of alternative (and more
+detailed) dBFT specification.
```
Comment (roman-khimov) [inline formal-models/README.md:95]: Say specifically that in theory these models are equivalent?

File: formal-models/dbft/dbft.tla [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+MoreThanFNodesCommitted(r) == CountCommitted(r) > F
+
+\* PrepareRequestSentOrReceived denotes whether there's a PrepareRequest
+\* message received from the current round's speaker (as the node r sees it).
+PrepareRequestSentOrReceived(r) == [type |-> "PrepareRequest", rm |-> GetPrimary(rmState[r].view), view |-> rmState[r].view] \in msgs
+
+\* -------------- Safety temporal formula --------------
+
+\* Init is the initial predicate initializing values at the start of every
+\* behaviour.
```
Comment (roman-khimov) [inline formal-models/dbft/dbft.tla:120]: check?

File: formal-models/dbft/dbft.tla [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+  /\ UNCHANGED <<>>
+
+\* RMSendPrepareResponse describes non-primary node r receiving PrepareRequest from
+\* the primary node of the current round (view) and broadcasting PrepareResponse.
+\* This step assumes that PrepareRequest always contains valid transactions and
+\* signatures.
+RMSendPrepareResponse(r) ==
+  /\ \/ rmState[r].type = "initialized"
+     \/ /\ rmState[r].type = "cv"
+        /\ MoreThanFNodesCommitted(r)
```
Comment (roman-khimov) [inline formal-models/dbft/dbft.tla:140]: This transition is one of most unobvious ones, probably deserves a comment (relates to NotAcceptingPayloadsDueToViewChanging logic).

**Comment by roman-khimov** [comment]
In general it's very descriptive and nice, but maybe we need to specifically mention `MoreThanFNodesCommittedOrLost` in that this logic can't be reliable in detecting lost nodes (even with neo-project/neo#2057), because real nodes can go down at any time.

File: formal-models/README.md [hunk 6]
```diff
@@ -0,0 +1,135 @@
+## dBFT formal models
+
+This section contains the set of dBFT's formal specifications written in
```
[the remark this answers is not in the document]
Comment (AnnaShaleva) [reply formal-models/README.md]: Fixed.

File: formal-models/README.md (hunk 2 above)
Comment (AnnaShaleva) [reply formal-models/README.md:37]: Yes, it's in the https://app.creately.com/d/cAEwPtsWZGV/edit, but I've just realize that its free plan is too restricted, so we probably need to choose some other plotting tool. Anyway, it was just a quick sketch, so it doesn't take a lot of time to draw it once again. I actually have the whole set of flow diagrams drawn in the notebook.

File: formal-models/README.md (hunk 3 above)
Comment (AnnaShaleva) [reply formal-models/README.md]: Fixed. Actually, TLA+ have a nice feature of *specification implementations* which perfectly suites for the case of Extended model, but I didn't do this.

File: formal-models/dbft/dbft.tla (hunk 5 above)
Comment (AnnaShaleva) [reply formal-models/dbft/dbft.tla:152]: Fixed. Do we need to write a couple of words about it in the README? I've added a comment in the model itself.

File: formal-models/README.md (hunk 2 above)
Comment (AnnaShaleva) [reply formal-models/README.md:37]: Here's the file: https://drive.google.com/file/d/1ugE3jXq5EkmFdI2r5bWRazzM5oZNoLze/view?usp=sharing
and here's the tool: https://app.diagrams.net

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**